### ISSUE TITLE
Usage and format questions

### ISSUE THREAD

**Issue by adrinjalali** [huggingface/doc-builder#133]
It seems we're migrating all repos to use this project. I have a few questions which I think would be nice to have answers to while we migrate (ideally I would have liked to have answers to them before we migrate):

Taken from https://github.com/huggingface/huggingface_hub/issues/759

By reading the ["how to write documentation"](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation), it's clear how to write docstrings, but it's not clear how to write user guides and examples:

- I'm thinking something along the lines of having something resembling `sphinx-gallery` examples, and then referencing those examples from user guides.
- How does one reference other objects, classes, and methods from the current library and external libraries that are rendered with a hyperlink to the corresponding API doc?
- how to we reference other sections in the user guide from the API doc and the user guide itself?
- what is the workflow to add directives such as `versionadded, versionchanged`, etc?
- I see this in the docs:

> It helps to keep the old links working when renaming section header and/or moving sections from one document to another.

In `rst` you reference a tag, so renaming headers doesn't actually do anything to break existing links. How does `mdx` handle this?

I'm not familiar with `mdx`, so I'm not sure how to approach these issues.

cc @LysandreJik

**Comment by LysandreJik** [comment]
> I'm thinking something along the lines of having something resembling sphinx-gallery examples, and then referencing those examples from user guides.

Do you mean you'd like to have a tool that from a .py file with docstring, generates a HTML render of that file and puts it in a given folder? Right now the tool takes everything and 

> How does one reference other objects, classes, and methods from the current library and external libraries that are rendered with a hyperlink to the corresponding API doc?

If I'm not misunderstanding, this is covered in the guide regarding how to write the documentation:

"_If you want to create a link to some internal class or function, you need to provide its path. For instance: [`file_utils.ModelOutput`]. This will be converted into a link with file_utils.ModelOutput in the description. To get rid of the path and only keep the name of the object you are linking to in the description, add a ~: [`~file_utils.ModelOutput`] will generate a link with ModelOutput in the description._"

The same holds true for cross-library references: `` [`~datasets.Dataset`] `` will do a link to the `Dataset` object documented in the `datasets` library.

> how to we reference other sections in the user guide from the API doc and the user guide itself?

As far as I know, other sections/parts of the documentation can be referenced by their local paths: `` [text](/other_page) `` and `` [test](/other_page#section) ``

> what is the workflow to add directives such as versionadded, versionchanged, etc?

For the beginning, I think we would put them manually as a `<Note>` so that it's displayed prominently on the methods. @mishig25, what do you think? See the [following directive in Sphinx](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded), it would be very useful for `huggingface_hub` as downstream libraries would be aware of what became available when.

> In rst you reference a tag, so renaming headers doesn't actually do anything to break existing links. How does mdx handle this?

I think @mishig25 has actually been involved heavily in the links/headers, so may be able to provide a better answer than I can.

**Comment by adrinjalali** [comment]
> Do you mean you'd like to have a tool that from a .py file with docstring, generates a HTML render of that file and puts it in a given folder? Right now the tool takes everything

I mean something like this [file](https://github.com/scikit-learn/scikit-learn/blob/main/examples/ensemble/plot_stack_predictors.py) which is then rendered like this [page](https://scikit-learn.org/stable/auto_examples/ensemble/plot_stack_predictors.html) which also generates a python notebook from the `.py` file and you can also run the notebook on binder.

**Comment by sgugger** [comment]
We already have the ability to auto-generate notebooks from all the doc files, it sounds like what you want to do is similar, we just do it from a .mdx instead of a .py.

See for instance the [quicktour](https://huggingface.co/docs/transformers/quicktour) of Transformers with the open in colab/AWS studio at the top.

**Comment by sgugger** [comment]
A few other comments: 

> The same holds true for cross-library references: [`~datasets.Dataset`] will do a link to the Dataset object documented in the datasets library.

@LysandreJik note that this feature has not been developed yet, but it's on the roadmap :-) 

We can definitely add support for the versionadded, versionchanged directives. We didn't use them in Transformers and Datasets, which is why they're not supported yet.

> In rst you reference a tag, so renaming headers doesn't actually do anything to break existing links. How does mdx handle this?

You can define an html anchor and reference it. See for instance the [Training tutorial in transformers](https://raw.githubusercontent.com/huggingface/transformers/master/docs/source/training.mdx). GitHub does not let me link to specific lines but you have a `trainer` tag at line 64, and you can then link to it via `training#trainer` from any page in the doc.

**Comment by adrinjalali** [comment]
Awesome, this clarifies things quite a bit for me.

> We already have the ability to auto-generate notebooks from all the doc files, it sounds like what you want to do is similar, we just do it from a .mdx instead of a .py.
> 
> See for instance the [quicktour](https://huggingface.co/docs/transformers/quicktour) of Transformers with the open in colab/AWS studio at the top.

This is pretty cool, and from the user's perspective, exactly what I was hoping to see. However, what I think is missing is from the development side: working on a code heavy documentation page where there is a lot of code, it's much easier to work on a `.py` file where the contributor/developer can test/run/debug the file and the notebook cells in the same file. For instance, consider this part from the quick tour file:

```` mdx
For a larger dataset where the inputs are big (like in speech or vision), you will 
want to pass along a generator instead of a list that loads all the inputs in memory. 
See the [pipeline documentation](./main_classes/pipelines) for more information.

### Use another model and tokenizer in the pipeline

The [`pipeline`] can accommodate any model from the [Model Hub](https://huggingface.co/models),
 making it easy to adapt the [`pipeline`] for other use-cases. For example, if you'd like 
a model capable of handling French text, use the tags on the Model Hub to filter for 
an appropriate model. The top filtered result returns a multilingual 
[BERT model](https://huggingface.co/nlptown/bert-base-multilingual-uncased-sentiment) 
fine-tuned for sentiment analysis. Great, let's use this model!

```py
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
```

Use the [`AutoModelForSequenceClassification`] and ['AutoTokenizer'] to load 
the pretrained model and it's associated tokenizer (more on an `AutoClass` below):

```py
>>> from transformers import AutoTokenizer, AutoModelForSequenceClassification

>>> model = AutoModelForSequenceClassification.from_pretrained(model_name)
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
>>> # ===PT-TF-SPLIT===
>>> from transformers import AutoTokenizer, TFAutoModelForSequenceClassification

>>> model = TFAutoModelForSequenceClassification.from_pretrained(model_name)
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
```
````

The same doc could potentially look like:

``` py
# %%
# For a larger dataset where the inputs are big (like in speech or vision), you
# will want to pass along a generator instead of a list that loads all the
# inputs in memory. See the [pipeline documentation](./main_classes/pipelines)
# for more information.

# %%
# Use another model and tokenizer in the pipeline
# -----------------------------------------------
#
# The [`pipeline`] can accommodate any model from the [Model
# Hub](https://huggingface.co/models), making it easy to adapt the [`pipeline`]
# for other use-cases. For example, if you'd like a model capable of handling
# French text, use the tags on the Model Hub to filter for an appropriate
# model. The top filtered result returns a multilingual [BERT
# model](https://huggingface.co/nlptown/bert-base-multilingual-uncased-sentiment)
# fine-tuned for sentiment analysis. Great, let's use this model!

model_name = "nlptown/bert-base-multilingual-uncased-sentiment"

# %% 
# Use the [`AutoModelForSequenceClassification`] and ['AutoTokenizer'] to load
# the pretrained model and it's associated tokenizer (more on an `AutoClass`
# below):

from transformers import AutoTokenizer, AutoModelForSequenceClassification

model = AutoModelForSequenceClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)
# ===PT-TF-SPLIT===
from transformers import AutoTokenizer, TFAutoModelForSequenceClassification

model = TFAutoModelForSequenceClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)
```
I don't think this is an urgent request/feature. But `sphinx-gallery` which handles these cases has been quite helpful in the projects I've worked in, and it'd be nice to have something similar for the doc-build infra we're building.

**Comment by sgugger** [comment]
Good point! We can definitely add support for this and it shouldn't be difficult to parse such a py file to generate a MDX and the corresponding notebooks.

The main difference with your suggestion is that the text would be in full Markdown, so not
```
# Use another model and tokenizer in the pipeline
# -----------------------------------------------
```
but
```
# ## Use another model and tokenizer in the pipeline
```
